File: app/src/core/regles.ts
/**
 * Frontière unique entre l'application et le cœur métier.
 *
 * Les règles du produit (classement de Maintenant, ordre de la Revue, ancrage dans le
 * texte source) vivent dans le module Kotlin `core/`, compilé en JS. Cette frontière
 * expose exactement les fonctions du pont `ZeNoteRegles` : entrée JSON, sortie
 * JSON, aucun état. Aucun autre fichier de l'application ne doit réimplémenter ces
 * règles ni contourner ce module.
 */

export interface ElementJson {
  id: string;
  captureId: string;
  type: 'TACHE' | 'ENGAGEMENT' | 'ATTENTE' | 'INFORMATION' | 'DECISION' | 'IDEE';
  texte: string;
  debutCar: number;
  finCar: number;
  debutMs?: number | null;
  finMs?: number | null;
  echeance?: string | null;
  echeanceConfiance?: number | null;
  echeanceIndice?: string | null;
  poids?: 'FAIBLE' | 'MOYEN' | 'FORT' | null;
  poidsConfiance?: number | null;
  poidsIndice?: string | null;
  interlocuteur?: string | null;
  interlocuteurConfiance?: number | null;
  sphere?: 'PROFESSIONNEL' | 'PERSONNEL' | null;
  planDeclencheur?: string | null;
  planAction?: string | null;
  verdict: 'EN_ATTENTE' | 'ACCEPTE' | 'UN_JOUR' | 'REJETE';
  corrigeParHumain: boolean;
}

export interface PropositionJson {
  elementId: string;
  texte: string;
  raison: string;
  poidsEffectif: string;
  urgence: string;
}

export interface EntreeRevueJson {
  element: ElementJson;
  aConfirmer: boolean;
  planManquant: boolean;
  urgence: string;
}

/** Une relance proposée en Revue : le produit se souvient à votre place. */
export interface RelanceJson {
  elementId: string;
  texte: string;
  /** ENGAGEMENT ou ATTENTE. */
  type: string;
  interlocuteur: string | null;
  echeance: string | null;
  /** Pourquoi elle remonte maintenant. Jamais un reproche. */
  motif: string;
  /** RELANCER, PROLONGER, CLORE. */
  options: string[];
}

/** La dernière nouvelle connue sur une attente. */
export interface SuiviJson {
  elementId: string;
  derniereNouvelle: string;
}

export interface RevueJson {
  groupes: { captureId: string; entrees: EntreeRevueJson[] }[];
  /** La file entière, y compris ce qui n'est pas présenté aujourd'hui. */
  total: number;
  /**
   * `true` quand la file dépassait ce qu'une Revue absorbe. Les groupes ne portent
   * alors que les entrées retenues : le reste demeure en file, intact.
   */
  reduite: boolean;
  /** La phrase du cœur expliquant la réduction. L'écran ne la réécrit pas. */
  motifReduction: string;
  /** Combien d'entrées demeurent en file. Un reste, jamais un retard. */
  demeurentEnFile: number;
}

export interface AncrageJson {
  retenus: ElementJson[];
  ecartes: { texte: string; raison: string }[];
}

/** Une capture telle qu'on la passe à la recherche : ce qu'elle lit, rien de plus. */
export interface CaptureJson {
  id: string;
  texte: string;
  creeLe: string;
  /**
   * Le jour de la capture vu par son auteur, en ISO (`AAAA-MM-JJ`).
   *
   * Distinct de `creeLe`, qui est en temps universel : une capture de 23 h 30 y tombe
   * le lendemain. Seule l'application sait dans quel fuseau son porteur vit. Absent,
   * la capture reste hors de portée des questions à repère temporel — elle n'est
   * jamais rattachée à une période au hasard.
   */
  jour?: string | null;
}

/** Ce que la surface retient d'un rappel entre deux ouvertures de l'application. */
export interface SuiviRappelJson {
  elementId: string;
  /** Le moment où le plan a été attaché, en heure locale (`AAAA-MM-JJTHH:MM`). */
  planPoseLe: string;
  /** Combien de fois ce rappel a déjà été présenté puis écarté sans être traité. */
  foisIgnore: number;
}

/** Un rappel présenté à un point de rupture. */
export interface RappelLivreJson {
  elementId: string;
  texte: string;
  /** Le signal tel que l'utilisateur l'a formulé. */
  declencheur: string;
  /**
   * Non vide quand ZeNote ne sait pas observer ce signal et l'a ramené à la reprise
   * de l'appareil. L'écran l'affiche : un rappel qui arrive au mauvais moment sans le
   * dire est pire qu'un rappel absent.
   */
  substitution: string;
  /** `true` si le signal s'était produit avant ce point de rupture. Un constat. */
  enRetard: boolean;
}

/** Un rappel qui ne se représente plus à l'identique, et que la Revue reprend. */
export interface EscaladeJson {
  elementId: string;
  texte: string;
  motif: string;
  /** REPLANIFIER, DELEGUER, ABANDONNER. */
  options: string[];
}

/**
 * Ce qu'un point de rupture livre. `titre` est vide quand il n'y a rien à présenter —
 * et il n'y a alors rien à afficher : une notification vide est une interruption sans
 * contenu.
 */
export interface RappelsDuMomentJson {
  titre: string;
  rappels: RappelLivreJson[];
  escalades: EscaladeJson[];
}

/** Ce dont une réponse de recherche se réclame. Jamais un score. */
export interface CitationJson {
  captureId: string;
  extrait: string;
  pourquoi: string;
  elementId: string | null;
}

/**
 * Une réponse de recherche. `fondee` vaut `false` quand rien ne correspond : l'énoncé
 * le dit alors, et `citations` est vide. C'est ce qui rend impossible d'afficher une
 * affirmation que rien ne porte.
 */
export interface ReponseJson {
  question: string;
  enonce: string;
  fondee: boolean;
  citations: CitationJson[];
  indisponibleHorsLigne: string[];
  /**
   * Ce que la question demandait et que le produit ne sait pas faire — par
   * construction, pas par panne. L'écran l'affiche : répondre à moitié sans le dire
   * laisserait croire que la question entière a été honorée.
   */
  nonPrisEnCompte: string[];
}

// ----------------------------------------------------------- le vrai cœur ---
// Les règles ci-dessous ne sont PAS écrites ici : elles viennent du module
// Kotlin `core/`, compilé en JavaScript et déposé dans `vendor/zenote-core/`
// par `scripts/sync-core-js.sh`.
//
// C'est délibéré et c'est le point d'architecture qui tient tout : la PWA
// d'aujourd'hui et les applications natives de demain appellent exactement le
// même classement. Réécrire ces règles en TypeScript les ferait diverger — ce
// que toute la conception cherche à empêcher.

import coeur from '../../vendor/zenote-core/zenote-core.js';

const Regles = (coeur as any).app.zenote.core.js.ZeNoteRegles as {
  maintenant(elementsJson: string, aujourdhui: string): string;
  revue(elementsJson: string, aujourdhui: string): string;
  transcriptionLisible(brut: string): string;
  filtrerAncrage(texteSource: string, elementsJson: string): string;
  relances(
    elementsJson: string,
    aujourdhui: string,
    suivisJson: string,
    delaisJson: string,
  ): string;
  rechercherParMots(
    requete: string,
    elementsJson: string,
    capturesJson: string,
    reseau: boolean,
  ): string;
  rechercherParQuestion(
    requete: string,
    elementsJson: string,
    capturesJson: string,
    aujourdhui: string,
    reseau: boolean,
  ): string;
  rappels(elementsJson: string, maintenant: string, suivisJson: string): string;
  rechercherParPersonne(personne: string, elementsJson: string, reseau: boolean): string;
  readonly version: string;
};

/** Version du contrat portée par le cœur : elle doit valoir celle attendue ici. */
export const VERSION_CONTRAT_ATTENDUE = '6';

if (Regles.version !== VERSION_CONTRAT_ATTENDUE) {
  throw new Error(
    `Cœur incompatible : contrat ${Regles.version} alors que l'application attend ` +
      `${VERSION_CONTRAT_ATTENDUE}. Relancez scripts/sync-core-js.sh.`,
  );
}

/** Seuil sous lequel une déduction devient une question posée en Revue. */
export const SEUIL_CONFIANCE = 0.75;

/** La vue Maintenant ne montre jamais plus de trois choses à la fois. */
export const MAX_PROPOSITIONS = 3;

/** Vue Maintenant : `ElementJson[]` + date ISO → `PropositionJson[]`. */
export function maintenant(elementsJson: string, aujourdhui: string): string {
  return Regles.maintenant(elementsJson, aujourdhui);
}

/** File de Revue : `ElementJson[]` + date ISO → `RevueJson`. */
export function revue(elementsJson: string, aujourdhui: string): string {
  return Regles.revue(elementsJson, aujourdhui);
}

/**
 * La version lisible d'une transcription : hésitations et répétitions immédiates en
 * moins, le sens intact.
 *
 * Rend le texte inchangé quand il n'y a rien à retirer : l'appelant sait alors qu'il
 * n'a pas deux versions à proposer. Le brut reste la couche source — c'est lui que
 * l'extraction lit, les ancrages étant des positions dans ce texte-là.
 */
export function transcriptionLisible(brut: string): string {
  return Regles.transcriptionLisible(brut);
}

/** Ancrage : texte source + `ElementJson[]` → `AncrageJson`. */
export function filtrerAncrage(texteSource: string, elementsJson: string): string {
  return Regles.filtrerAncrage(texteSource, elementsJson);
}

/** Relances du jour : `ElementJson[]` + `SuiviJson[]` + délais → `RelanceJson[]`. */
export function relances(
  elementsJson: string,
  aujourdhui: string,
  suivisJson: string,
  delaisJson: string,
): string {
  return Regles.relances(elementsJson, aujourdhui, suivisJson, delaisJson);
}

/** Recherche par mots : `ElementJson[]` + `CaptureJson[]` → `ReponseJson`. */
export function rechercherParMots(
  requete: string,
  elementsJson: string,
  capturesJson: string,
  reseau: boolean,
): string {
  return Regles.rechercherParMots(requete, elementsJson, capturesJson, reseau);
}

/**
 * Recherche par question : les mots, plus le repère temporel qu'elle porte —
 * « la semaine dernière », « avant-hier », « il y a trois jours ».
 */
export function rechercherParQuestion(
  requete: string,
  elementsJson: string,
  capturesJson: string,
  aujourdhui: string,
  reseau: boolean,
): string {
  return Regles.rechercherParQuestion(requete, elementsJson, capturesJson, aujourdhui, reseau);
}

/**
 * Rappels d'un point de rupture : `ElementJson[]` + date-heure locale +
 * `SuiviRappelJson[]` → `RappelsDuMomentJson`.
 */
export function rappels(
  elementsJson: string,
  maintenant: string,
  suivisJson: string,
): string {
  return Regles.rappels(elementsJson, maintenant, suivisJson);
}

/** Recherche par personne : nom + `ElementJson[]` → `ReponseJson`. */
export function rechercherParPersonne(
  personne: string,
  elementsJson: string,
  reseau: boolean,
): string {
  return Regles.rechercherParPersonne(personne, elementsJson, reseau);
}

// ------------------------------------------------------- confort d'appel ----
// Enveloppes typées : elles ne contiennent aucune règle, seulement le (dé)codage
// JSON du contrat ci-dessus. Elles restent valables après le branchement Kotlin.

export function maintenantObjets(elements: ElementJson[], aujourdhui: string): PropositionJson[] {
  return JSON.parse(maintenant(JSON.stringify(elements), aujourdhui)) as PropositionJson[];
}

export function revueObjets(elements: ElementJson[], aujourdhui: string): RevueJson {
  return JSON.parse(revue(JSON.stringify(elements), aujourdhui)) as RevueJson;
}

export function filtrerAncrageObjets(texteSource: string, elements: ElementJson[]): AncrageJson {
  return JSON.parse(filtrerAncrage(texteSource, JSON.stringify(elements))) as AncrageJson;
}

export function rechercherParMotsObjets(
  requete: string,
  elements: ElementJson[],
  captures: CaptureJson[],
  reseau: boolean,
): ReponseJson {
  return JSON.parse(
    rechercherParMots(requete, JSON.stringify(elements), JSON.stringify(captures), reseau),
  ) as ReponseJson;
}

export function rechercherParQuestionObjets(
  requete: string,
  elements: ElementJson[],
  captures: CaptureJson[],
  aujourdhui: string,
  reseau: boolean,
): ReponseJson {
  return JSON.parse(
    rechercherParQuestion(
      requete,
      JSON.stringify(elements),
      JSON.stringify(captures),
      aujourdhui,
      reseau,
    ),
  ) as ReponseJson;
}

export function rappelsObjets(
  elements: ElementJson[],
  maintenant: string,
  suivis: SuiviRappelJson[],
): RappelsDuMomentJson {
  return JSON.parse(
    rappels(JSON.stringify(elements), maintenant, JSON.stringify(suivis)),
  ) as RappelsDuMomentJson;
}

export function rechercherParPersonneObjets(
  personne: string,
  elements: ElementJson[],
  reseau: boolean,
): ReponseJson {
  return JSON.parse(
    rechercherParPersonne(personne, JSON.stringify(elements), reseau),
  ) as ReponseJson;
}

export function relancesObjets(
  elements: ElementJson[],
  aujourdhui: string,
  suivis: SuiviJson[] = [],
  delais: Record<string, number> = {},
): RelanceJson[] {
  return JSON.parse(
    relances(JSON.stringify(elements), aujourdhui, JSON.stringify(suivis), JSON.stringify(delais)),
  ) as RelanceJson[];
}

File: core/src/commonMain/kotlin/app/zenote/core/api/Regles.kt
package app.zenote.core.api

import app.zenote.core.model.CaptureId
import app.zenote.core.model.Deduit
import app.zenote.core.model.ElementDerive
import app.zenote.core.model.ElementId
import app.zenote.core.model.ElementResolu
import app.zenote.core.model.Passage
import app.zenote.core.model.Plan
import app.zenote.core.model.Poids
import app.zenote.core.model.Sphere
import app.zenote.core.model.TypeElement
import app.zenote.core.model.Verdict
import app.zenote.core.priorisation.ContexteMaintenant
import app.zenote.core.priorisation.Priorisation
import app.zenote.core.rappels.Declencheur
import app.zenote.core.rappels.Echeance
import app.zenote.core.rappels.Echeancier
import app.zenote.core.rappels.FileOpportunite
import app.zenote.core.rappels.Livraison
import app.zenote.core.rappels.PointDeRupture
import app.zenote.core.rappels.Rappel
import app.zenote.core.rappels.RappelId
import app.zenote.core.recherche.RechercheLocale
import app.zenote.core.revue.Arriere
import app.zenote.core.revue.FileRevue
import app.zenote.core.revue.Relance
import app.zenote.core.revue.Suivi
import app.zenote.core.recherche.Reponse
import app.zenote.core.recherche.TexteSource
import kotlinx.datetime.LocalDate
import kotlinx.datetime.LocalDateTime
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toInstant
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json

/**
 * Les règles du produit, exposées aux surfaces par un contrat JSON.
 *
 * Le cœur est **sans état** vu d'ici : la surface détient les données et les passe à
 * chaque appel. C'est ce qui permet à la PWA de tout garder en local, et à une
 * application native de faire le même appel sur les mêmes règles.
 */
object Regles {

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    /**
     * La vue Maintenant : au plus trois éléments, chacun avec sa justification.
     *
     * @param elementsJson tableau d'[ElementJson]
     * @param aujourdhui date ISO `AAAA-MM-JJ`
     * @return tableau de [PropositionJson]
     */
    fun maintenant(elementsJson: String, aujourdhui: String): String {
        val elements = decoder(elementsJson)
        val propositions = Priorisation.maintenant(
            elements = elements.map { it.versResolu() },
            contexte = ContexteMaintenant(LocalDate.parse(aujourdhui)),
        ).map { p ->
            PropositionJson(
                elementId = p.element.id.value,
                texte = p.element.texte,
                raison = p.raison,
                poidsEffectif = p.poidsEffectif.name,
                urgence = p.urgence.name,
            )
        }
        return json.encodeToString(ListSerializer(PropositionJson.serializer()), propositions)
    }

    /**
     * La file de Revue : ce qui attend une décision, groupé par capture source pour
     * que l'utilisateur n'ait à relire la source qu'une fois, et ordonné pour que
     * l'urgent et l'incertain passent devant.
     *
     * @return un [RevueJson]
     */
    fun revue(elementsJson: String, aujourdhui: String): String {
        val date = LocalDate.parse(aujourdhui)
        val enAttente = decoder(elementsJson).filter { it.verdict == Verdict.EN_ATTENTE.name }

        // La file passe d'abord par le domaine : c'est lui qui sait ce qu'une Revue
        // absorbe, et ce qu'il vaut mieux laisser en file plutôt que de le déverser.
        val parId = enAttente.associateBy { it.id }
        val entrees = enAttente.map { FileRevue.entree(it.versResolu(), date) }
        val reduction = Arriere.revueReduite(entrees)

        val retenues = reduction.retenues.map { entree ->
            val dto = parId.getValue(entree.id.value)
            EntreeRevueJson(
                element = dto,
                aConfirmer = entree.aConfirmer,
                planManquant = entree.planAFournir,
                urgence = entree.urgence.name,
            )
        }

        // Les captures sont présentées dans l'ordre de leur entrée la plus pressante ;
        // à l'intérieur d'un groupe, l'urgent puis l'incertain d'abord.
        val groupes = retenues
            .groupBy { it.element.captureId }
            .map { (captureId, dansLeGroupe) ->
                GroupeRevueJson(
                    captureId = captureId,
                    entrees = dansLeGroupe.sortedWith(
                        compareBy<EntreeRevueJson> { urgenceOrdinale(it.urgence) }
                            .thenByDescending { it.aConfirmer }
                            .thenBy { it.element.id },
                    ),
                )
            }
            .sortedWith(
                compareBy<GroupeRevueJson> { groupe ->
                    groupe.entrees.minOf { urgenceOrdinale(it.urgence) }
                }.thenBy { it.captureId },
            )

        return json.encodeToString(
            RevueJson.serializer(),
            RevueJson(
                groupes = groupes,
                total = entrees.size,
                reduite = reduction.reduite,
                motifReduction = if (reduction.reduite) reduction.motif else "",
                demeurentEnFile = reduction.demeurentEnFile.size,
            ),
        )
    }

    /**
     * Ce que la Revue du jour doit remonter d'elle-même : un engagement dont
     * l'échéance approche, une attente restée sans nouvelle au-delà du délai habituel
     * de la personne concernée.
     *
     * C'est la moitié du produit que l'utilisateur ne peut pas réclamer, puisqu'il l'a
     * précisément oubliée.
     *
     * @param suivisJson tableau de [SuiviJson] — la dernière nouvelle connue par attente
     * @param delaisJson objet JSON `{ "personne": jours }`, délais habituels observés
     * @return tableau de [RelanceJson]
     */
    fun relances(
        elementsJson: String,
        aujourdhui: String,
        suivisJson: String,
        delaisJson: String,
    ): String {
        val suivis = json
            .decodeFromString(ListSerializer(SuiviJson.serializer()), suivisJson)
            .map { Suivi(ElementId(it.elementId), LocalDate.parse(it.derniereNouvelle)) }
        val delais = json
            .decodeFromString(MapSerializer(String.serializer(), Int.serializer()), delaisJson)

        val propositions = Relance.aRelancer(
            elements = decoder(elementsJson).map { it.versResolu() },
            aujourdhui = LocalDate.parse(aujourdhui),
            suivis = suivis,
            delaisObserves = delais,
        ).map { p ->
            RelanceJson(
                elementId = p.element.id.value,
                texte = p.element.texte,
                type = p.element.type.name,
                interlocuteur = p.element.interlocuteur,
                echeance = p.element.echeance?.toString(),
                motif = p.motif,
                options = p.options.map { it.name },
            )
        }

        return json.encodeToString(ListSerializer(RelanceJson.serializer()), propositions)
    }

    /**
     * Ce qu'un point de rupture doit présenter, et ce qui remonte en Revue.
     *
     * Le produit n'a qu'un point de rupture observable depuis un navigateur : la
     * reprise de l'appareil, c'est-à-dire le retour dans l'application après une
     * pause. Fin de réunion et fin de créneau demandent l'agenda, qui n'est pas
     * branché — ils viendront avec lui, sans rien changer ici.
     *
     * Trois règles sont tenues par [FileOpportunite], et pas réécrites ici : une
     * seule notification par point de rupture, les rappels groupés dedans, et
     * l'escalade en Revue au bout de trois fois ignoré. Les compteurs d'ignorés sont
     * rejoués depuis ce que la surface a retenu — la file est une machine à états
     * pure, et la rejouer donne exactement l'état où on l'avait laissée.
     *
     * @param elementsJson tableau d'[ElementJson]
     * @param maintenant date-heure locale `AAAA-MM-JJTHH:MM`
     * @param suivisJson tableau de [SuiviRappelJson]
     * @return un [RappelsDuMomentJson]
     */
    fun rappels(elementsJson: String, maintenant: String, suivisJson: String): String {
        val instant = LocalDateTime.parse(maintenant)
        // La file raisonne en `Instant` ; elle ne fait que comparer les siens entre
        // eux. Les lire tous dans le même fuseau suffit donc, et évite de faire entrer
        // une question de fuseau là où il n'y en a pas.
        val a = instant.toInstant(TimeZone.UTC)

        val suivis = json
            .decodeFromString(ListSerializer(SuiviRappelJson.serializer()), suivisJson)
            .associateBy { it.elementId }

        // Seuls les éléments acceptés qui portent un plan sont des rappels : le plan
        // est ce qui transforme une note en quelque chose qui doit revenir.
        val candidats = decoder(elementsJson)
            .map { it.versResolu() }
            .filter { it.verdict == Verdict.ACCEPTE && it.plan != null }
            .sortedBy { it.id.value }

        val file = FileOpportunite()
        val rappels = candidats.map { element ->
            element to Rappel(
                id = RappelId(element.id.value),
                elementId = element.id,
                texte = element.plan!!.action.ifBlank { element.texte },
                declencheur = Declencheur.Transition(PointDeRupture.REPRISE_APPAREIL),
            )
        }

        // Rejouer les ignorés d'abord : un rappel déjà escaladé doit l'être à nouveau
        // avant qu'on tente de le déposer, sinon il repasserait une fois de trop.
        for ((element, rappel) in rappels) {
            repeat(suivis[element.id.value]?.foisIgnore ?: 0) { file.ignorer(rappel) }
        }

        val substitutions = mutableMapOf<String, String>()
        val retards = mutableSetOf<String>()
        for ((element, rappel) in rappels) {
            val suivi = suivis[element.id.value] ?: continue
            val echeance = Echeancier.quand(
                declencheur = element.plan!!.declencheur,
                poseLe = LocalDateTime.parse(suivi.planPoseLe),
            )
            if (!Echeancier.estArrive(echeance, instant)) continue

            when (echeance) {
                is Echeance.Substituee -> substitutions[element.id.value] = echeance.explication
                is Echeance.Observable -> if (echeance.quand < instant) retards += element.id.value
            }
            file.deposer(rappel, a)
        }

        val notification = file.vider(PointDeRupture.REPRISE_APPAREIL, a)
        val parId = candidats.associateBy { it.id.value }

        return json.encodeToString(
            RappelsDuMomentJson.serializer(),
            RappelsDuMomentJson(
                titre = notification?.titre ?: "",
                rappels = notification?.rappels.orEmpty().map { r ->
                    RappelLivreJson(
                        elementId = r.elementId.value,
                        texte = r.texte,
                        declencheur = parId[r.elementId.value]?.plan?.declencheur ?: "",
                        substitution = substitutions[r.elementId.value] ?: "",
                        enRetard = r.elementId.value in retards,
                    )
                },
                escalades = file.escalades().map { e ->
                    EscaladeJson(
                        elementId = e.rappel.elementId.value,
                        texte = e.rappel.texte,
                        motif = e.motif,
                        options = e.options.map { it.name },
                    )
                },
            ),
        )
    }

    /**
     * L'ancrage : ne garde que les éléments dont le passage se relit vraiment dans le
     * texte source. C'est le filet entre le modèle et l'écran.
     *
     * Les éléments écartés sont rendus avec leur raison, pour être traçables plutôt
     * que disparaître en silence.
     *
     * @return un [AncrageJson]
     */
    fun filtrerAncrage(texteSource: String, elementsJson: String): String {
        val retenus = mutableListOf<ElementJson>()
        val ecartes = mutableListOf<EcarteJson>()

        decoder(elementsJson).forEach { dto ->
            val raison = raisonDeRejet(dto, texteSource)
            if (raison == null) retenus += dto else ecartes += EcarteJson(dto.texte, raison)
        }

        return json.encodeToString(
            AncrageJson.serializer(),
            AncrageJson(retenus = retenus, ecartes = ecartes),
        )
    }

    /**
     * Recherche par mots, sur les éléments et sur le texte des captures pas encore
     * structurées. Rien ne sort de l'appareil : c'est une lecture de ce qui est là.
     *
     * @param capturesJson tableau de [CaptureJson] — les captures encore non analysées
     *   comptent, sinon la recherche mentirait par omission à qui vient de dicter
     * @param reseau `false` en mode avion : les résultats locaux sont rendus quand même,
     *   et ce qui est en pause est signalé plutôt que masqué
     * @return un [ReponseJson]
     */
    fun rechercherParMots(
        requete: String,
        elementsJson: String,
        capturesJson: String,
        reseau: Boolean,
    ): String = rendre(
        RechercheLocale.parMots(
            requete = requete,
            elements = decoder(elementsJson).map { it.versResolu() },
            captures = sources(capturesJson),
            reseau = reseau,
        ),
    )

    /**
     * Recherche par question : les mots, plus le repère temporel qu'elle porte —
     * « la semaine dernière », « avant-hier », « il y a trois jours ».
     *
     * @return un [ReponseJson]
     */
    fun rechercherParQuestion(
        requete: String,
        elementsJson: String,
        capturesJson: String,
        aujourdhui: String,
        reseau: Boolean,
    ): String = rendre(
        RechercheLocale.parQuestion(
            requete = requete,
            elements = decoder(elementsJson).map { it.versResolu() },
            captures = sources(capturesJson),
            aujourdhui = LocalDate.parse(aujourdhui),
            reseau = reseau,
        ),
    )

    /** Les captures telles que la recherche les lit. */
    private fun sources(capturesJson: String): List<TexteSource> = json
        .decodeFromString(ListSerializer(CaptureJson.serializer()), capturesJson)
        .map {
            TexteSource(
                captureId = CaptureId(it.id),
                texte = it.texte,
                quand = it.creeLe,
                // Une date que la surface n'a pas fournie n'est pas devinée depuis
                // l'horodatage : elle serait juste la plupart du temps, et fausse le
                // soir, ce qui est la pire des combinaisons pour une recherche.
                jour = it.jour?.let(LocalDate::parse),
            )
        }

    /**
     * Recherche par personne : ce qui a été promis à quelqu'un, et ce qu'on attend
     * d'elle. Les éléments clos sont rendus aussi, en le disant.
     *
     * @return un [ReponseJson]
     */
    fun rechercherParPersonne(
        personne: String,
        elementsJson: String,
        reseau: Boolean,
    ): String = rendre(
        RechercheLocale.parPersonne(
            personne = personne,
            elements = decoder(elementsJson).map { it.versResolu() },
            reseau = reseau,
        ),
    )

    // ------------------------------------------------------------------ interne

    private fun rendre(reponse: Reponse): String = json.encodeToString(
        ReponseJson.serializer(),
        ReponseJson(
            question = reponse.question,
            enonce = reponse.enonce,
            fondee = reponse.fondee,
            citations = reponse.citations.map {
                CitationJson(
                    captureId = it.captureId.value,
                    extrait = it.extrait,
                    pourquoi = it.pourquoi,
                    elementId = it.elementId?.value,
                )
            },
            indisponibleHorsLigne = reponse.indisponibleHorsLigne,
            nonPrisEnCompte = reponse.nonPrisEnCompte,
        ),
    )


    private fun decoder(elementsJson: String): List<ElementJson> =
        json.decodeFromString(ListSerializer(ElementJson.serializer()), elementsJson)

    private fun raisonDeRejet(dto: ElementJson, texteSource: String): String? = when {
        dto.texte.isBlank() -> "élément sans texte"
        dto.debutCar < 0 || dto.finCar <= dto.debutCar -> "passage source vide ou incohérent"
        dto.finCar > texteSource.length -> "passage source absent du texte de la capture"
        runCatching { TypeElement.valueOf(dto.type) }.isFailure -> "type inconnu : ${dto.type}"
        else -> null
    }

    private fun urgenceOrdinale(nom: String): Int =
        runCatching { app.zenote.core.priorisation.Urgence.valueOf(nom).ordinal }
            .getOrDefault(Int.MAX_VALUE)

    private fun ElementJson.aConfirmer(): Boolean = listOfNotNull(
        echeanceConfiance,
        poidsConfiance,
        interlocuteurConfiance,
    ).any { it < app.zenote.core.model.SEUIL_CONFIANCE }

    /** Traduit le format de fil vers le domaine, décisions humaines déjà appliquées. */
    private fun ElementJson.versResolu(): ElementResolu {
        val derive = ElementDerive(
            captureId = CaptureId(captureId),
            type = TypeElement.valueOf(type),
            texte = texte,
            passage = Passage(debutCar, finCar, debutMs, finMs),
            echeance = echeance?.let {
                Deduit(LocalDate.parse(it), echeanceConfiance ?: 1.0, echeanceIndice ?: "fourni")
            },
            poids = poids?.let {
                Deduit(Poids.valueOf(it), poidsConfiance ?: 1.0, poidsIndice ?: "fourni")
            },
            interlocuteur = interlocuteur?.let {
                Deduit(it, interlocuteurConfiance ?: 1.0, "nommé")
            },
            sphere = sphere?.let { Deduit(Sphere.valueOf(it), 1.0, "déduit") },
            plan = planDeclencheur?.let { d ->
                planAction?.let { a -> Deduit(Plan(d, a), 1.0, "plan posé en Revue") }
            },
        )
        return ElementResolu(
            id = ElementId(id),
            captureId = derive.captureId,
            type = derive.type,
            texte = derive.texte,
            passage = derive.passage,
            echeance = derive.echeance?.valeur,
            poids = derive.poids?.valeur,
            interlocuteur = derive.interlocuteur?.valeur,
            sphere = derive.sphere?.valeur,
            plan = derive.plan?.valeur,
            verdict = Verdict.valueOf(verdict),
            aConfirmer = aConfirmer(),
            corrigeParHumain = corrigeParHumain,
            indicePoids = if (corrigeParHumain && poids != null) "poids fixé à la main" else poidsIndice,
        )
    }
}
